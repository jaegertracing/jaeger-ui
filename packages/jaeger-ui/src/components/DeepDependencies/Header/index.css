/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgHeader {
  align-items: stretch;
  background: #fafafa;
  border-bottom: 1px solid #ddd;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  position: relative;
  z-index: 10;
}

.DdgHeader--controlHeader {
  align-items: center;
  background-color: #f6f6f6;
  display: flex;
  justify-content: space-between;
  padding-left: 1.25rem;
}

.DdgHeader--paramsHeader {
  align-items: center;
  background: #fafafa;
  border-bottom: 1px solid #eee;
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  font-size: unset;
  margin: 0;
  padding: 0.75rem 1.25rem 0.75rem 1.25rem;
  position: relative;
}

.DdgHeader--findWrapper {
  display: flex;
  flex: 1 1 auto;
  justify-content: flex-end;
}

.DdgHeader--uiFind {
  align-items: center;
  background-color: #fff;
  border-left: 1px solid #eee;
  border-right: 1px solid #eee;
  color: #bbb;
  display: flex;
  flex-grow: 1;
  max-width: 400px;
  min-width: 200px;
  transition: max-width 0.5s;
}

.DdgHeader--uiFind:focus-within {
  max-width: 550px;
}

.DdgHeader--uiFindSearchIcon {
  margin: 0 0.25em;
  padding: 0.24em;
}

.DdgHeader--uiFindInput > input:focus,
.DdgHeader--uiFindInput > input {
  border: none;
  box-shadow: none;
  border-radius: 0px;
}

.DdgHeader--uiFindInfo {
  background: #ffffb8;
  border: 1px dashed #fadb14;
  border-radius: 4px;
  color: #ad8b00;
  margin-left: 0.25em;
  margin-right: 0.5em;
  padding: 0.27em 0.4em;
  stroke: #666;
  stroke-width: 1.2px;
  text-align: center;
  transition: width 0.5s;
  white-space: nowrap;
}

.DdgHeader--uiFindInfo--hidden {
  border-left: dashed var(--tx-color-muted) 1px;
  margin-left: 0.3em;
  padding-left: 0.3em;
}

.DdgHeader--uiFindInfo--icon {
  margin-bottom: 1.5px;
  margin-left: 0.2em;
}

.DdgHeader--uiFindInfo--tooltip {
  white-space: nowrap;
}
