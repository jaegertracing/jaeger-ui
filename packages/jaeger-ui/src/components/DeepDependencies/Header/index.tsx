// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Icon, Input, Tooltip } from 'antd';
import MdVisibility from 'react-icons/lib/md/visibility';
import MdVisibilityOff from 'react-icons/lib/md/visibility-off';

import HopsSelector from './HopsSelector';
import NameSelector from '../../common/NameSelector';
import LayoutSettings from './LayoutSettings';
import { trackFilter, trackHeaderSetOperation, trackShowMatches } from '../index.track';
import UiFindInput from '../../common/UiFindInput';
import { EDirection, TDdgDistanceToPathElems, EDdgDensity } from '../../../model/ddg/types';

import './index.css';

type TProps = {
  clearOperation: () => void;
  density: EDdgDensity;
  distanceToPathElems?: TDdgDistanceToPathElems;
  hiddenUiFindMatches?: Set<string>;
  operation?: string;
  operations: string[] | undefined;
  service?: string;
  services?: string[] | null;
  setDensity: (density: EDdgDensity) => void;
  setDistance: (distance: number, direction: EDirection) => void;
  setOperation: (operation: string) => void;
  setService: (service: string) => void;
  showOperations: boolean;
  showParameters?: boolean;
  showVertices: (vertexKeys: string[]) => void;
  toggleShowOperations: (enable: boolean) => void;
  uiFindCount: number | undefined;
  visEncoding?: string;
};
export default class Header extends React.PureComponent<TProps> {
  private _uiFindInput: React.RefObject<Input> = React.createRef();

  static defaultProps = {
    showParameters: true,
  };

  focusUiFindInput = () => {
    if (this._uiFindInput.current) {
      this._uiFindInput.current.focus();
    }
  };

  getUiFindInfo = () => {
    const { hiddenUiFindMatches, uiFindCount } = this.props;

    if (uiFindCount === undefined) return null;

    let hasHidden = false;
    let hiddenInfo: React.ReactNode = null;
    let tipText = uiFindCount ? 'All matches are visible' : 'No matches';
    if (hiddenUiFindMatches && hiddenUiFindMatches.size) {
      const { size } = hiddenUiFindMatches;
      hasHidden = true;
      tipText = `Click to view ${size} hidden match${size !== 1 ? 'es' : ''}`;
      hiddenInfo = (
        <span className="DdgHeader--uiFindInfo--hidden">
          {size}
          <MdVisibilityOff className="DdgHeader--uiFindInfo--icon" />
        </span>
      );
    }

    return (
      <Tooltip overlayClassName="DdgHeader--uiFindInfo--tooltip" placement="topRight" title={tipText}>
        {/* arbitrary span is necessary as Tooltip alters child's styling */}
        <span>
          <button
            className="DdgHeader--uiFindInfo"
            disabled={!hasHidden}
            onClick={this.handleInfoClick}
            type="button"
          >
            {uiFindCount}
            {(uiFindCount !== 0 || hasHidden) && <MdVisibility className="DdgHeader--uiFindInfo--icon" />}
            {hiddenInfo}
          </button>
        </span>
      </Tooltip>
    );
  };

  setOperation = (operation: string) => {
    trackHeaderSetOperation();
    this.props.setOperation(operation);
  };

  handleInfoClick = () => {
    trackShowMatches();
    const { hiddenUiFindMatches, showVertices } = this.props;
    if (hiddenUiFindMatches) showVertices(Array.from(hiddenUiFindMatches));
  };

  render() {
    const {
      clearOperation,
      density,
      distanceToPathElems,
      operation,
      operations,
      service,
      services,
      setDensity,
      setDistance,
      setService,
      showOperations,
      showParameters,
      toggleShowOperations,
      visEncoding,
    } = this.props;

    return (
      <header className="DdgHeader">
        {showParameters && (
          <div className="DdgHeader--paramsHeader">
            <NameSelector
              label="Service"
              placeholder="Select a service…"
              value={service || null}
              setValue={setService}
              required
              options={services || []}
            />
            {service && (
              <NameSelector
                clearValue={clearOperation}
                label="Operation"
                placeholder="Filter by operation…"
                value={operation || null}
                setValue={this.setOperation}
                options={operations || []}
              />
            )}
          </div>
        )}
        <div className="DdgHeader--controlHeader">
          <LayoutSettings
            density={density}
            setDensity={setDensity}
            showOperations={showOperations}
            toggleShowOperations={toggleShowOperations}
          />
          <HopsSelector
            distanceToPathElems={distanceToPathElems}
            handleClick={setDistance}
            visEncoding={visEncoding}
          />
          <div className="DdgHeader--findWrapper">
            <div className="DdgHeader--uiFind" role="button" onClick={this.focusUiFindInput}>
              <Icon className="DdgHeader--uiFindSearchIcon" type="search" />
              <UiFindInput
                allowClear
                forwardedRef={this._uiFindInput}
                inputProps={{ className: 'DdgHeader--uiFindInput' }}
                trackFindFunction={trackFilter}
              />
              {this.getUiFindInfo()}
            </div>
          </div>
        </div>
      </header>
    );
  }
}
