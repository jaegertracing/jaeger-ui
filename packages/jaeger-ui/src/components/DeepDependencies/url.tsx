// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import _isEmpty from 'lodash/isEmpty';
import memoizeOne from 'memoize-one';
import queryString from 'query-string';
import { matchPath } from 'react-router-dom';

import { EDdgDensity, TDdgSparseUrlState } from '../../model/ddg/types';
import prefixUrl from '../../utils/prefix-url';

export const ROUTE_PATH = prefixUrl('/deep-dependencies');

const ROUTE_MATCHER = { path: ROUTE_PATH, strict: true, exact: true };

export function matches(path: string) {
  return Boolean(matchPath(path, ROUTE_MATCHER));
}

export function getUrl(args?: { [key: string]: unknown; showOp?: boolean }, baseUrl: string = ROUTE_PATH) {
  if (args && !_isEmpty(args)) {
    const stringifyArgs =
      Reflect.has(args, 'showOp') && args.showOp !== undefined
        ? {
            ...args,
            showOp: args.showOp ? 1 : 0,
          }
        : args;
    return `${baseUrl}?${queryString.stringify(stringifyArgs)}`;
  }
  return baseUrl;
}

function firstParam(arg: string | string[]): string {
  if (Array.isArray(arg)) {
    const returnVal = arg[0];
    console.warn(`Found multiple query parameters: "${arg}", using "${returnVal}"`); // eslint-disable-line no-console
    return returnVal;
  }
  return arg;
}

export const getUrlState = memoizeOne(function getUrlState(search: string): TDdgSparseUrlState {
  const {
    density = EDdgDensity.PreventPathEntanglement,
    decoration,
    end,
    hash,
    operation,
    service,
    showOp,
    start,
    visEncoding,
  } = queryString.parse(search);
  const rv: TDdgSparseUrlState = {
    density: firstParam(density) as EDdgDensity,
  };
  if (decoration) {
    rv.decoration = firstParam(decoration);
  }
  if (end) {
    rv.end = Number.parseInt(firstParam(end), 10);
  }
  if (hash) {
    rv.hash = firstParam(hash);
  }
  if (operation) {
    rv.operation = firstParam(operation);
  }
  if (service) {
    rv.service = firstParam(service);
  }
  if (showOp) {
    rv.showOp = Boolean(+firstParam(showOp));
  }
  if (start) {
    rv.start = Number.parseInt(firstParam(start), 10);
  }
  if (visEncoding) {
    rv.visEncoding = firstParam(visEncoding);
  }
  return rv;
});

export const sanitizeUrlState = memoizeOne(function sanitizeUrlStateImpl(
  state: TDdgSparseUrlState,
  hash?: string
): TDdgSparseUrlState {
  if (hash && state.hash === hash) {
    return state;
  }
  const { visEncoding, ...sanitized } = state; // eslint-disable-line @typescript-eslint/no-unused-vars
  return sanitized;
});
