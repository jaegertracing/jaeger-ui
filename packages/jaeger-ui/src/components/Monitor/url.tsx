// Copyright (c) 2021 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { matchPath } from 'react-router-dom';

import prefixUrl from '../../utils/prefix-url';

export const ROUTE_PATH = prefixUrl('/monitor');

const ROUTE_MATCHER = { path: ROUTE_PATH, strict: true, exact: true };

export function matches(path: string) {
  return Boolean(matchPath(path, ROUTE_MATCHER));
}

export function getUrl() {
  return ROUTE_PATH;
}
