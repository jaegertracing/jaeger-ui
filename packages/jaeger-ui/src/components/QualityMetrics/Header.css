/*
Copyright (c) 2020 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.QualityMetrics--Header {
  align-items: center;
  background: #fafafa;
  border-bottom: 1px solid #ddd;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: row;
  padding: 0.75rem 1.25rem 0.75rem 1.25rem;
  position: relative;
  z-index: 10;
}

.QualityMetrics--Header--LookbackLabel {
  color: var(--tx-color-muted);
  cursor: pointer;
  font-size: 1.5em;
  margin: 0 0.6em 0 0;
  outline: 4px solid transparent;
}

.QualityMetrics--Header--LookbackSuffix {
  color: var(--tx-color-muted);
  margin: 0 0 0 0.6em;
  outline: 4px solid transparent;
}
