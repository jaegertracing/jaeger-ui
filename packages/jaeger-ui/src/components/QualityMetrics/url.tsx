// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import memoizeOne from 'memoize-one';
import queryString from 'query-string';
import { matchPath } from 'react-router-dom';

import prefixUrl from '../../utils/prefix-url';

export const ROUTE_PATH = prefixUrl('/quality-metrics');

const ROUTE_MATCHER = { path: ROUTE_PATH, strict: true, exact: true };

export function matches(path: string) {
  return Boolean(matchPath(path, ROUTE_MATCHER));
}

export function getUrl(queryParams?: Record<string, string | number>) {
  if (!queryParams) return ROUTE_PATH;

  return `${ROUTE_PATH}?${queryString.stringify(queryParams)}`;
}

type TReturnValue = {
  lookback: number;
  service?: string;
};

export const getUrlState = memoizeOne(function getUrlState(search: string): TReturnValue {
  const { lookback: lookbackFromUrl, service: serviceFromUrl } = queryString.parse(search);
  const service = Array.isArray(serviceFromUrl) ? serviceFromUrl[0] : serviceFromUrl;
  const lookbackStr = Array.isArray(lookbackFromUrl) ? lookbackFromUrl[0] : lookbackFromUrl;
  const lookback = lookbackStr && Number.parseInt(lookbackStr, 10);
  const rv: TReturnValue = {
    lookback: 48,
  };
  if (service) rv.service = service;
  if (lookback) rv.lookback = lookback;
  return rv;
});
