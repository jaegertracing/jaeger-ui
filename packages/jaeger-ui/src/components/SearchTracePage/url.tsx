// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import queryString from 'query-string';
import { matchPath } from 'react-router-dom';

import prefixUrl from '../../utils/prefix-url';

import { SearchQuery } from '../../types/search';

function eqEq(a: string | number | null | undefined, b: string | number | null | undefined) {
  return (a == null && b == null) || String(a) === String(b);
}

export const ROUTE_PATH = prefixUrl('/search');

const ROUTE_MATCHER = { path: ROUTE_PATH, strict: true, exact: true };

export function matches(path: string) {
  return Boolean(matchPath(path, ROUTE_MATCHER));
}

export function getUrl(query?: Record<string, unknown> | null | undefined) {
  const search = query ? `?${queryString.stringify(query)}` : '';
  return prefixUrl(`/search${search}`);
}

export function isSameQuery(a: SearchQuery, b: SearchQuery) {
  if (Boolean(a) !== Boolean(b)) {
    return false;
  }
  return (
    eqEq(a.end, b.end) &&
    eqEq(a.limit, b.limit) &&
    eqEq(a.lookback, b.lookback) &&
    eqEq(a.maxDuration, b.maxDuration) &&
    eqEq(a.minDuration, b.minDuration) &&
    eqEq(a.operation, b.operation) &&
    eqEq(a.service, b.service) &&
    eqEq(a.start, b.start) &&
    eqEq(a.tags, b.tags)
  );
}
