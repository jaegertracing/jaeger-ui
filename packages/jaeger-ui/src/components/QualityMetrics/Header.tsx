// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { InputNumber } from 'antd';
import _debounce from 'lodash/debounce';

import NameSelector from '../common/NameSelector';

import './Header.css';

type TProps = {
  lookback: number;
  service?: string;
  services?: string[] | null;
  setLookback: (lookback: number | string | undefined) => void;
  setService: (service: string) => void;
};

type TState = {
  ownInputValue: number | undefined;
};

export default class Header extends React.PureComponent<TProps, TState> {
  state: TState = {
    ownInputValue: undefined,
  };

  setLookback = _debounce((lookback: number | string | undefined) => {
    this.setState({ ownInputValue: undefined });
    this.props.setLookback(lookback);
  }, 350);

  handleInputChange = (value: string | number | undefined) => {
    if (typeof value === 'string') return;
    this.setState({ ownInputValue: value });
    this.setLookback(value);
  };

  render() {
    const { lookback, service, services, setService } = this.props;
    const { ownInputValue } = this.state;
    const lookbackValue = ownInputValue !== undefined ? ownInputValue : lookback;

    return (
      <header className="QualityMetrics--Header">
        <NameSelector
          label="Service"
          placeholder="Select a serviceâ€¦"
          value={service || null}
          setValue={setService}
          required
          options={services || []}
        />
        <label className="QualityMetrics--Header--LookbackLabel" htmlFor="inputNumber">
          Lookback:
        </label>
        <InputNumber id="inputNumber" onChange={this.handleInputChange} min={1} value={lookbackValue} />
        <span className="QualityMetrics--Header--LookbackSuffix">(in hours)</span>
      </header>
    );
  }
}
