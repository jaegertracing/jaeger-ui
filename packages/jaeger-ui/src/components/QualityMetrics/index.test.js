// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';

import JaegerAPI from '../../api/jaeger';
import Header from './Header';
import * as getUrl from './url';
import { UnconnectedQualityMetrics, mapDispatchToProps, mapStateToProps } from '.';

describe('QualityMetrics', () => {
  describe('UnconnectedQualityMetrics', () => {
    const props = {
      fetchServices: jest.fn(),
      history: {
        push: jest.fn(),
      },
      lookback: 48,
      service: 'test-service',
      services: ['foo', 'bar', 'baz'],
    };
    const { service: _s, ...propsWithoutService } = props;
    let fetchQualityMetricsSpy;
    let promise;
    let res;
    let rej;

    beforeAll(() => {
      fetchQualityMetricsSpy = jest.spyOn(JaegerAPI, 'fetchQualityMetrics').mockImplementation(() => {
        promise = new Promise((resolve, reject) => {
          res = resolve;
          rej = reject;
        });
        return promise;
      });
    });

    beforeEach(() => {
      props.history.push.mockClear();
      props.fetchServices.mockClear();
      fetchQualityMetricsSpy.mockClear();
    });

    describe('constructor', () => {
      it('fetches services if none are provided', () => {
        const { services: _ses, ...propsWithoutServices } = props;
        // eslint-disable-next-line no-new
        new UnconnectedQualityMetrics(propsWithoutServices);
        expect(props.fetchServices).toHaveBeenCalledTimes(1);
      });

      it('no-ops if services are provided', () => {
        // eslint-disable-next-line no-new
        new UnconnectedQualityMetrics(props);
        expect(props.fetchServices).not.toHaveBeenCalled();
      });
    });

    describe('componentDidMount', () => {
      it('fetches quality metrics', () => {
        shallow(<UnconnectedQualityMetrics {...props} />);
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(1);
        expect(fetchQualityMetricsSpy).toHaveBeenCalledWith(props.service, props.lookback);
      });
    });

    describe('componentDidUpdate', () => {
      const expectedState = {
        qualityMetrics: undefined,
        error: undefined,
        loading: true,
      };
      const initialState = {
        qualityMetrics: {
          scores: [],
          metrics: [],
        },
        error: {},
        loading: false,
      };
      let wrapper;

      beforeEach(() => {
        wrapper = shallow(<UnconnectedQualityMetrics {...props} />);
        wrapper.setState(initialState);
      });

      it('clears state and fetches quality metrics if service changed', () => {
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(1);

        const service = `not-${props.service}`;
        wrapper.setProps({ service });
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(2);
        expect(fetchQualityMetricsSpy).toHaveBeenLastCalledWith(service, props.lookback);
        expect(wrapper.state()).toEqual(expectedState);
      });

      it('clears state and fetches quality metrics if lookback changed', () => {
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(1);

        const lookback = `not-${props.lookback}`;
        wrapper.setProps({ lookback });
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(2);
        expect(fetchQualityMetricsSpy).toHaveBeenLastCalledWith(props.service, lookback);
        expect(wrapper.state()).toEqual(expectedState);
      });

      it('no-ops if neither service or lookback changed', () => {
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(1);

        wrapper.setProps({ services: [] });
        expect(fetchQualityMetricsSpy).toHaveBeenCalledTimes(1);
      });
    });

    describe('fetches quality metrics', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = shallow(<UnconnectedQualityMetrics {...propsWithoutService} />);
      });

      it('no-ops on falsy service', () => {
        expect(wrapper.state('loading')).toBe(undefined);
      });

      it('fetches quality metrics and updates state on success', async () => {
        wrapper.setProps({ service: props.service });
        expect(wrapper.state('loading')).toBe(true);

        const qualityMetrics = {};
        res(qualityMetrics);
        await promise;

        expect(wrapper.state('loading')).toBe(false);
        expect(wrapper.state('qualityMetrics')).toBe(qualityMetrics);
      });

      it('fetches quality metrics and updates state on error', async () => {
        wrapper.setProps({ service: props.service });
        expect(wrapper.state('loading')).toBe(true);

        const error = {};
        rej(error);
        await promise.catch(() => {});

        expect(wrapper.state('loading')).toBe(false);
        expect(wrapper.state('error')).toBe(error);
      });
    });

    describe('url updates', () => {
      const testLookback = props.lookback * 4;
      const testUrl = 'test.url';
      let getUrlSpy;
      let latestUrl;
      let setLookback;
      let wrapper;

      beforeAll(() => {
        getUrlSpy = jest.spyOn(getUrl, 'getUrl').mockImplementation(() => {
          latestUrl = `${testUrl}.${getUrlSpy.mock.calls.length}`;
          return latestUrl;
        });
        wrapper = shallow(<UnconnectedQualityMetrics {...props} />);
        setLookback = wrapper.find(Header).prop('setLookback');
      });

      it('sets service', () => {
        const testService = `new ${props.service}`;
        wrapper.find(Header).prop('setService')(testService);
        expect(getUrlSpy).toHaveBeenLastCalledWith({
          lookback: props.lookback,
          service: testService,
        });
        expect(props.history.push).toHaveBeenLastCalledWith(latestUrl);
      });

      it('sets lookback', () => {
        setLookback(testLookback);
        expect(getUrlSpy).toHaveBeenLastCalledWith({
          lookback: testLookback,
          service: props.service,
        });
        expect(props.history.push).toHaveBeenLastCalledWith(latestUrl);
      });

      it('sets lookback without service', () => {
        wrapper.setProps({ service: undefined });
        setLookback(testLookback);
        expect(getUrlSpy).toHaveBeenLastCalledWith({
          lookback: testLookback,
          service: '',
        });
        expect(props.history.push).toHaveBeenLastCalledWith(latestUrl);
      });

      it('ignores falsy lookback', () => {
        setLookback(null);
        expect(props.history.push).not.toHaveBeenCalled();
      });

      it('ignores string lookback', () => {
        setLookback(props.service);
        expect(props.history.push).not.toHaveBeenCalled();
      });

      it('ignores less than one lookback', () => {
        setLookback(-1);
        expect(props.history.push).not.toHaveBeenCalled();
      });

      it('ignores fractional lookback', () => {
        setLookback(2.5);
        expect(props.history.push).not.toHaveBeenCalled();
      });
    });

    describe('render', () => {
      it('renders without loading, error, or metrics', () => {
        expect(shallow(<UnconnectedQualityMetrics {...propsWithoutService} />)).toMatchSnapshot();
      });

      it('renders when loading', () => {
        expect(shallow(<UnconnectedQualityMetrics {...props} />)).toMatchSnapshot();
      });

      it('renders when errored', async () => {
        const wrapper = shallow(<UnconnectedQualityMetrics {...props} />);
        const message = 'Error message';
        rej({ message });
        await promise.catch(() => {});
        expect(wrapper).toMatchSnapshot();
      });

      it('renders with metrics', async () => {
        const wrapper = shallow(<UnconnectedQualityMetrics {...props} />);
        const metrics = {
          bannerText: 'test banner text',
          traceQualityDocumentationLink: 'trace.quality.documentation/link',
          scores: [
            {
              key: 'score0',
            },
            {
              key: 'score1',
            },
          ],
          metrics: [
            {
              name: 'metric 0',
            },
            {
              name: 'metric 1',
            },
          ],
        };
        res(metrics);
        await promise;
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('mapDispatchToProps()', () => {
    it('creates the actions correctly', () => {
      expect(mapDispatchToProps(() => {})).toEqual({
        fetchServices: expect.any(Function),
      });
    });
  });

  describe('mapStateToProps()', () => {
    let getUrlStateSpy;
    const urlState = {
      lookback: 108,
      service: 'test-service',
    };

    beforeAll(() => {
      getUrlStateSpy = jest.spyOn(getUrl, 'getUrlState').mockImplementation(search => search && urlState);
    });

    it('gets services from redux state', () => {
      const services = [urlState.service, 'foo', 'bar'];
      expect(mapStateToProps({ services: { services } }, { location: {} })).toEqual({ services });
    });

    it('gets current service and lookback from url', () => {
      const search = 'test search';
      expect(mapStateToProps({ services: {} }, { location: { search } })).toEqual(urlState);
      expect(getUrlStateSpy).toHaveBeenLastCalledWith(search);
    });
  });
});
