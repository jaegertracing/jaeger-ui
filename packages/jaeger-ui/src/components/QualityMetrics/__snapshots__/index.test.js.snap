// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QualityMetrics UnconnectedQualityMetrics render renders when errored 1`] = `
<div
  className="QualityMetrics"
>
  <Header
    lookback={48}
    service="test-service"
    services={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    setLookback={[Function]}
    setService={[Function]}
  />
  <div
    className="QualityMetrics--Error"
  >
    Error message
  </div>
</div>
`;

exports[`QualityMetrics UnconnectedQualityMetrics render renders when loading 1`] = `
<div
  className="QualityMetrics"
>
  <Header
    lookback={48}
    service="test-service"
    services={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    setLookback={[Function]}
    setService={[Function]}
  />
  <LoadingIndicator
    centered={true}
    small={false}
  />
</div>
`;

exports[`QualityMetrics UnconnectedQualityMetrics render renders with metrics 1`] = `
<div
  className="QualityMetrics"
>
  <Header
    lookback={48}
    service="test-service"
    services={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    setLookback={[Function]}
    setService={[Function]}
  />
  <BannerText
    bannerText="test banner text"
  />
  <div
    className="QualityMetrics--Body"
  >
    <div
      className="QualityMetrics--ScoreCards"
    >
      <ScoreCard
        key="score0"
        link="trace.quality.documentation/link"
        score={
          Object {
            "key": "score0",
          }
        }
      />
      <ScoreCard
        key="score1"
        link="trace.quality.documentation/link"
        score={
          Object {
            "key": "score1",
          }
        }
      />
    </div>
    <div
      className="QualityMetrics--MetricCards"
    >
      <MetricCard
        key="metric 0"
        metric={
          Object {
            "name": "metric 0",
          }
        }
      />
      <MetricCard
        key="metric 1"
        metric={
          Object {
            "name": "metric 1",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`QualityMetrics UnconnectedQualityMetrics render renders without loading, error, or metrics 1`] = `
<div
  className="QualityMetrics"
>
  <Header
    lookback={48}
    services={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    setLookback={[Function]}
    setService={[Function]}
  />
</div>
`;
